source/common/router/config_impl.h

```diff
@@ -658,8 +658,22 @@ class RouteEntryImplBase : public RouteEntryAndRoute,
     return opaque_config_;
   }
   bool includeVirtualHostRateLimits() const override { return include_vh_rate_limits_; }
-  const envoy::config::core::v3::Metadata& metadata() const override { return metadata_; }
-  const Envoy::Config::TypedMetadata& typedMetadata() const override { return typed_metadata_; }
+  using DefaultMetadata = ConstSingleton<envoy::config::core::v3::Metadata>;
+  const envoy::config::core::v3::Metadata& metadata() const override {
+    if (metadata_ != nullptr) {
+      return *metadata_;
+    }
+    return DefaultMetadata::get();
+  }
+  using RouteTypedMetadata = Envoy::Config::TypedMetadataImpl<HttpRouteTypedMetadataFactory>;
+  const Envoy::Config::TypedMetadata& typedMetadata() const override {
+    if (typed_metadata_ != nullptr) {
+      return *typed_metadata_;
+    }
+    static RouteTypedMetadata* defaultTypedMetadata =
```

Comment:
nit: const.

Reply:
On a second thought, could this be a ConstSingleton by itself (i.e., not dependent on `DefaultMetadata::get()`)?
I'm trading off another object of empty Metadata object against dependencies between singletons. WDYT?

Reply:
ConstSingleton as defined does not allow constructors that need input arguments.

Reply:
What @pradeepcrao said, typed metadata needs a raw metadata instance to be created.

Reply:
Might be possible to do using type traits which will make this cleaner. I can attempt that in a follow-up PR.

Reply:
Added const. Leaving the singleton as is.